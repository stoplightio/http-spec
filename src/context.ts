import { hasRef, isLocalRef } from '@stoplight/json';
// @ts-ignore
import * as fnv from 'fnv-plus';

import type {
  AvailableContext,
  Fragment,
  IdGenerator,
  RefResolver,
  TransformerContext,
  TranslateFunction,
} from './types';

export const DEFAULT_ID_GENERATOR: IdGenerator = input => fnv.fast1a32hex(input);

export function createContext<T extends Record<string, unknown>>(
  document: T,
  resolveRef: RefResolver<T>,
  generateId: IdGenerator,
): TransformerContext<T> {
  let context: AvailableContext = 'service';
  return {
    document,
    get context() {
      return context;
    },
    set context(value) {
      context = value;
      if (value !== 'operation') {
        this.parentId = this.ids[value];
      }
    },
    maybeResolveLocalRef(target: unknown) {
      if (hasRef(target) && isLocalRef(target.$ref)) {
        try {
          return resolveRef.call(this, target);
        } catch {
          return target;
        }
      }

      return target;
    },
    generateId(value) {
      this.parentId = generateId(value);
      return this.parentId;
    },
    ids: {
      service: '',
      path: '',
      operation: '',
    },
    parentId: '',
  };
}

export function withContext<F extends TranslateFunction<any, any[]> = TranslateFunction<Fragment, unknown[]>>(
  fn: F,
): F {
  return <F>function (...args: Parameters<F>) {
    const { context, parentId } = this;
    const result = fn.apply(this, args);
    this.context = context;
    this.parentId = parentId;
    return result;
  };
}
