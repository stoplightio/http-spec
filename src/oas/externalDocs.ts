import { IExternalDocs } from '@stoplight/types';

/**
 * Extract an OpanAPI `externalDocs` object
 * @return an object that includes an `externalDocs` object with all well-formed
 * properties from the supplied object, or an empty object
 */
export function toExternalDocs(externalDocs: any): { externalDocs?: IExternalDocs } {
  if (!externalDocs || typeof externalDocs !== 'object') {
    return {};
  }

  const url = getUrl(externalDocs);
  if (url === undefined) {
    return {};
  }

  const description = getDescription(externalDocs);
  const descriptionProp = description ? { description } : {};

  return {
    externalDocs: {
      url,
      ...descriptionProp,
    },
  };
}

const ALL_WHITESPACE: RegExp = /^\s*$/;

/**
 * @returns the `url` property, or undefined if the object lacks it
 */
function getUrl(externalDocs: { url?: any } | undefined): string | undefined {
  const url = externalDocs?.url;
  if (!url || typeof url !== 'string' || ALL_WHITESPACE.test(url)) {
    return undefined;
  }

  return url;
}

/**
 * @returns the `description` property, or undefined if the object lacks it
 */
function getDescription(externalDocs: { description?: any }): string | undefined {
  const description = externalDocs.description;
  if (!description || typeof description !== 'string' || ALL_WHITESPACE.test(description)) {
    return undefined;
  }

  return description;
}
