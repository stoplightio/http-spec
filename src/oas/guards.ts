import { DeepPartial, Dictionary } from '@stoplight/types';
import { InfoObject } from 'openapi3-ts/src/model/OpenApi';
import { Info } from 'swagger-schema-official';

import { isDictionary } from '../utils';

export function hasXLogo(
  info: DeepPartial<Info | InfoObject>,
): info is DeepPartial<Info | InfoObject> & { 'x-logo': Dictionary<unknown> } {
  return isDictionary(info['x-logo']);
}
