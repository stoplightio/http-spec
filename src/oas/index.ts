export * from './operation';
export * from './types';
