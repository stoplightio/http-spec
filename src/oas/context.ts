// @ts-expect-error: no types
import * as fnv from 'fnv-plus';

import { createContext as _createContext } from '../context';
import type { Fragment, IdGenerator } from '../types';
import { resolveRef } from './resolver';

const HASH_32_KINDS = ['http_operation', 'schema'];

function getKind(input: string) {
  const i = input.indexOf('-');
  return input.slice(0, i);
}

const DEFAULT_ID_GENERATOR: IdGenerator = input => {
  const kind = getKind(input);
  return HASH_32_KINDS.some(k => kind === k) ? fnv.fast1a32hex(input) : fnv.fast1a52hex(input);
};

export function createContext<T extends Fragment>(document: T) {
  return _createContext<T>(document, resolveRef, DEFAULT_ID_GENERATOR);
}
