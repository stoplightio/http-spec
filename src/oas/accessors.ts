import type { Extensions } from '@stoplight/types';

import { Fragment, TransformerContext } from '../types';
import { entries } from '../utils';
import { isValidOas2Param, isValidOas3Param, Oas2ParamBase, Oas3ParamBase, ParamBase } from './guards';
import { OasVersion } from './types';

const ROOT_EXTENSIONS = ['x-internal'];

const getIdForParameter = (param: ParamBase) => `${param.name}-${param.in}`;

export function iterateOasParams(
  this: TransformerContext,
  spec: OasVersion.OAS2,
  path: Fragment,
  operation: Fragment,
): Iterable<Oas2ParamBase>;
export function iterateOasParams(
  this: TransformerContext,
  spec: OasVersion.OAS3,
  path: Fragment,
  operation: Fragment,
): Iterable<Oas3ParamBase>;
export function* iterateOasParams(
  this: TransformerContext,
  spec: OasVersion,
  path: Fragment,
  operation: Fragment,
): Iterable<Oas2ParamBase | Oas3ParamBase> {
  const seenParams = new Set();
  const { parentId, context } = this;
  const opParams = Array.isArray(operation.parameters) ? operation.parameters : [];
  const params = [...opParams, ...(Array.isArray(path.parameters) ? path.parameters : [])];

  for (let i = 0; i < params.length; i++) {
    const param = this.maybeResolveLocalRef(params[i]);
    if (!(spec === OasVersion.OAS2 ? isValidOas2Param : isValidOas3Param)(param)) continue;

    const key = getIdForParameter(param);

    if (seenParams.has(key)) continue;
    seenParams.add(key);

    if (this.context !== 'service') {
      this.context = i < opParams.length ? 'operation' : 'path';
    }

    yield param;
  }

  this.context = context;
  this.parentId = parentId;
}

export function getExtensions(target: unknown): Extensions {
  return Object.fromEntries(entries(target).filter(([key]) => key.startsWith('x-') && !ROOT_EXTENSIONS.includes(key)));
}
