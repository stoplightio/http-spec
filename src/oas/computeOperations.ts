import { IHttpOperation } from '@stoplight/types';
import { flatten, get, keys, map } from 'lodash';
import { OpenAPIObject } from 'openapi3-ts';
import { Spec } from 'swagger-schema-official';

import { HttpOperationTransformer } from '../oas/types';

const methods = ['get', 'post', 'put', 'delete', 'options', 'head', 'patch', 'trace'];

export function computeOasOperations(
  document: Spec | OpenAPIObject,
  transformer: HttpOperationTransformer<any>,
): IHttpOperation[] {
  const paths = keys(get(document, 'paths'));

  return flatten(
    map(paths, path =>
      keys(get(document, ['paths', path]))
        .filter(pathKey => methods.includes(pathKey))
        .map(method =>
          transformer({
            document,
            path,
            method,
          }),
        ),
    ),
  );
}
