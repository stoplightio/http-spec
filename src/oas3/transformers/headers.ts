import { isPlainObject } from '@stoplight/json';
import { HttpParamStyles, IHttpEncoding, IHttpHeaderParam, Optional, Reference } from '@stoplight/types';
import pickBy = require('lodash.pickby');

import { withContext } from '../../context';
import { isBoolean, isNonNullable, isString } from '../../guards';
import { isReferenceObject } from '../../oas/guards';
import { translateSchemaObject } from '../../oas/transformers';
import { translateToDefaultExample } from '../../oas/transformers/examples';
import { entries } from '../../utils';
import { isHeaderObject } from '../guards';
import { Oas3TranslateFunction } from '../types';
import { translateToExample } from './examples';

export const translateHeaderObject = withContext<
  Oas3TranslateFunction<
    [[name: string, headerObject: unknown]],
    Optional<IHttpHeaderParam<true> | (Pick<IHttpHeaderParam<true>, 'name'> & Reference)>
  >
>(function ([name, unresolvedHeaderObject]) {
  const maybeHeaderObject = this.maybeResolveLocalRef(unresolvedHeaderObject);

  if (isReferenceObject(maybeHeaderObject)) {
    (maybeHeaderObject as Pick<IHttpHeaderParam<true>, 'name'> & Reference).name = name;
    return maybeHeaderObject as Pick<IHttpHeaderParam<true>, 'name'> & Reference;
  }

  if (!isPlainObject(maybeHeaderObject)) return;

  const id = this.generateId.httpHeader({ nameOrKey: name });

  if (!isHeaderObject(maybeHeaderObject)) {
    return {
      id,
      encodings: [],
      examples: [],
      name,
      style: HttpParamStyles.Simple,
    };
  }

  const { content: contentObject } = maybeHeaderObject;

  const contentValue = isPlainObject(contentObject) ? Object.values(contentObject)[0] : null;

  const baseContent: IHttpHeaderParam = {
    id,
    name,
    style: HttpParamStyles.Simple,

    ...pickBy(
      {
        schema: isPlainObject(maybeHeaderObject.schema)
          ? translateSchemaObject.call(this, maybeHeaderObject.schema)
          : null,
        content: maybeHeaderObject.content,
      },
      isNonNullable,
    ),

    ...pickBy(
      {
        description: maybeHeaderObject.description,
      },
      isString,
    ),

    ...pickBy(
      {
        allowEmptyValue: maybeHeaderObject.allowEmptyValue,
        allowReserved: maybeHeaderObject.allowReserved,
        explode: maybeHeaderObject.explode,
        required: maybeHeaderObject.required,
        deprecated: maybeHeaderObject.deprecated,
      },
      isBoolean,
    ),
  };

  const examples: IHttpHeaderParam<true>['examples'] = [];
  const encodings: IHttpEncoding<true>[] = [];

  if (isPlainObject(contentValue)) {
    examples.push(...entries(contentValue.examples).map(translateToExample, this).filter(isNonNullable));

    if (isPlainObject(contentValue.encoding)) {
      encodings.push(...(Object.values(contentValue.encoding) as IHttpEncoding<true>[]));
    }

    if ('example' in contentValue) {
      examples.push(translateToDefaultExample.call(this, '__default_content', contentValue.example));
    }
  }

  examples.push(...entries(maybeHeaderObject.examples).map(translateToExample, this).filter(isNonNullable));

  if ('example' in maybeHeaderObject) {
    examples.push(translateToDefaultExample.call(this, '__default', maybeHeaderObject.example));
  }

  return {
    ...baseContent,
    encodings,
    examples,
  };
});
