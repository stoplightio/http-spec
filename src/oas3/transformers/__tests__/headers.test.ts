import { createContext } from '../../../oas/context';
import { translateHeaderObject as _translateHeaderObject } from '../headers';

const translateHeaderObject = (object: unknown, key: string) =>
  _translateHeaderObject.call(createContext({}), [key, object]);

describe('translateHeaderObject', () => {
  it('should translate to IHttpHeaderParam', () => {
    expect(
      translateHeaderObject(
        {
          description: 'descr',
          required: true,
          deprecated: true,
          allowEmptyValue: true,
          style: 'matrix',
          explode: true,
          allowReserved: true,
          schema: {},
          examples: {
            a: {
              summary: 'example summary',
              value: 'hey',
            },
            b: {
              summary: 'example summary',
              externalValue: 'https://stoplight.io/b',
            },
          },
          example: {},
          content: {},
        },
        'header-name',
      ),
    ).toMatchSnapshot({
      id: expect.any(String),
      examples: [
        {
          id: expect.any(String),
        },
        {
          id: expect.any(String),
        },
        {
          id: expect.any(String),
        },
      ],
      schema: {
        'x-stoplight': {
          id: expect.any(String),
        },
      },
    });
  });

  it('should handle nullish value gracefully', () => {
    expect(translateHeaderObject(null, 'header')).toBeUndefined();
  });
});
