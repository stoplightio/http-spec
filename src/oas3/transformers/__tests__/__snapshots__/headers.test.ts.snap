// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`translateHeaderObject should translate to IHttpHeaderParam 1`] = `
Object {
  "allowEmptyValue": true,
  "allowReserved": true,
  "content": Object {},
  "deprecated": true,
  "description": "descr",
  "encodings": Array [],
  "examples": Array [
    Object {
      "id": Any<String>,
      "key": "a",
      "summary": "example summary",
      "value": "hey",
    },
    Object {
      "externalValue": "https://stoplight.io/b",
      "id": Any<String>,
      "key": "b",
      "summary": "example summary",
    },
    Object {
      "id": Any<String>,
      "key": "__default",
      "value": Object {},
    },
  ],
  "explode": true,
  "id": Any<String>,
  "name": "header-name",
  "required": true,
  "schema": Object {
    "$schema": "http://json-schema.org/draft-07/schema#",
    "x-stoplight": Object {
      "id": Any<String>,
    },
  },
  "style": "simple",
}
`;
